feat: LaunchTemplate support for managed node-groups

philicious opened the pull request:
# PR o'clock

## Description

fixes #979

Just recently on 17th August, AWS released LaunchTemplate support for managed node-groups. https://aws.amazon.com/blogs/containers/introducing-launch-template-and-custom-ami-support-in-amazon-eks-managed-node-groups/
Furthermore the `aws` provider also supports it since `3.3.0` https://github.com/terraform-providers/terraform-provider-aws/pull/14639

This module didn't support it yet, only LTs for self-managed worker groups.

As the module is quite complex already, I only added support for providing the id of LT you create yourself and then supply the Id.
The existing `workers_launch_template.tf` couldn't have been easily reused imho as its related to also creating ASGs and other resources for self-managed node-groups. Also at least the `iam_instance_profile` should NOT be supplied for LTs being used for managed node-groups as I noticed. AWS API will error then.
So instead of adding another LT manifest and variables and wiring that all together, I prefered to take care of the LT myself and copied the userdata template 

I now create a LT like
```
data "template_file" "launch_template_userdata" {
  template = file("${path.module}/templates/userdata.sh.tpl")

  vars = {
    cluster_name        = var.cluster_name
    endpoint            = module.eks.cluster_endpoint
    cluster_auth_base64 = module.eks.cluster_certificate_authority_data

    bootstrap_extra_args = ""
    kubelet_extra_args = ""
    }
}

// this is mostly the default LT that AWS would create if you dont specify your own
resource "aws_launch_template" "default" {
  name_prefix     = "${var.cluster_name}-"
  description     = "Default Launch-Template for clusters"
  update_default_version = true

  block_device_mappings {
    device_name = "/dev/xvda"

    ebs {
      volume_size           = 100
      volume_type           = "gp2"
      delete_on_termination = true
    }
  }

  ebs_optimized = true // some instance types dont support it, so check when changing type

  image_id      = "ami-00341e507eb458a09" //TODO use our custom AMI
  instance_type = var.instance_type

  monitoring {
    enabled = true
  }

  network_interfaces {
    associate_public_ip_address = false
    delete_on_termination       = true
    security_groups             = [module.eks.worker_security_group_id]
  }

  user_data = base64encode(
    data.template_file.launch_template_userdata.rendered,
  )

  lifecycle {
    create_before_destroy = true
  }  
}
```

and pass it to the eks module
```
module "eks" {
  source          = "../../../../../terraform-aws-eks/"
  cluster_name    = var.cluster_name
  cluster_version = "1.16"
  subnets         = data.terraform_remote_state.network.outputs.private_subnets

  vpc_id = data.terraform_remote_state.network.outputs.vpc_id

  node_groups = {
    initial_group = {
      desired_capacity = 1
      max_capacity     = 15 // lets cap a cluster to 15 nodes, so ASGs cannot go insane
      min_capacity     = 1

      launch_template_id = aws_launch_template.default.id
      launch_template_version = aws_launch_template.default.default_version 

    }
  }  
...
...
}
```

Noteworthy: I had to wrap the userdata script in MIME as otherwise API complained it wasnt MIME. so I took the MIME wrapper as seen when creating a node group manually and letting AWS create a LT for you
```
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="//"

--//
Content-Type: text/x-shellscript; charset="us-ascii"
#!/bin/bash 
set -xe

# Bootstrap and join the cluster
/etc/eks/bootstrap.sh --b64-cluster-ca '${cluster_auth_base64}' --apiserver-endpoint '${endpoint}' ${bootstrap_extra_args} --kubelet-extra-args "${kubelet_extra_args}" '${cluster_name}'

--//--
```
### Checklist

- [x] CI tests are passing
- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

philicious commented:
PR would be ready. Maybe adding another example would help tho, how to actually use the `launch_template_id` as outlined above

calvinbui commented:
I had success using https://registry.terraform.io/providers/hashicorp/cloudinit/latest/docs/data-sources/cloudinit_config to wrap the userdata but I would say it shouldn't be part of this module like how you're making the LT not part of this module

calvinbui commented:
Though I must add that there should be some documentation with an example of what the LT should look like. 

i.e. instance profile cannot be set, EKS will complain about that

surajbarkale-dolby commented:
One good candidate for this is installing ssm agent on startup. See https://github.com/aws/containers-roadmap/issues/593#issuecomment-577181345

philicious commented:
I will add an example and make this PR fully ready towards weekend. 
I am super busy and stressed with customers roadmap atm

dpiddockcmp commented:
Does this actually work? Reading through [the documentation](https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html) they list some node group configurations that are prohibited when using launch templates:

| Prohibited value | Module default | aws provider default |
| ---- | ---- | ---- |
| ami\_type (if you specify custom AMI) | unset | `AL2_x86_64` |
| instance_type | `m4.large` | `t3.medium` |
| disk_size | unset | 20 GB |
| ssh keypair and extra security groups | unset | unset |

Or maybe by "prohibited" they mean ignored from the API values?

philicious commented:
Yes it does. Have cluster running with custom AMI by that

philicious commented:
@dpiddockcmp the documentation reads 
>The following table lists the settings that are prohibited in a managed node group configuration and which similar settings (if any) are required in a launch template.

so it is correct to **not** set the `disk_size` / `ami_type` and `instance_type`in the `node_groups` config block but rather in the `aws_launch_template` , as seen in my example :)

nxf5025 commented:
@dpiddockcmp 
> Does this actually work?

I tested this over the weekend and it worked great. Only addition I needed was to add `iam_role_arn` to the node_groups map:
```
  node_groups = {
    initial_group = {
      desired_capacity = 1
      max_capacity     = 15
      min_capacity     = 1
      iam_role_arn     = <arn>

      launch_template_id = aws_launch_template.default.id
      launch_template_version = aws_launch_template.default.default_version 

    }
  }  
```

karlderkaefer commented:
@philicious I can confirm that it works as you propose.

Carles-Figuerola commented:
Made a PR to make the random naming solution work properly when the node group needs to be replaced:
https://github.com/Bahn-X/terraform-aws-eks/pull/1

myoung34 commented:
Anything holding this back? I'd love to make use of it so we can use security groups and tags

philicious commented:
That PR of @Carles-Figuerola looks very good and besides that only example missing. Iam in holidays and had some stressful weeks at my client before but next week Wednesday could finish it up ! 
Using this productively at the client already :)

barryib commented:
Thanks @philicious for working on this. We have an terraform-aws-modules working session this friday. We'll discuss about the direction we want to take about this feature. We'll come back to you pretty soon.

huguesalary commented:
If that helps, I have tested this PR as well and it works as advertised, thank you @philicious and the maintainers of this repo for your work!

davidalger commented:
> When using a custom AMI, Amazon EKS doesn't merge any user data. Rather, you are responsible for supplying the required bootstrap commands for nodes to join the cluster. If your nodes fail to join the cluster, the Amazon EKS CreateNodegroup and UpdateNodegroupVersion actions also fail. 

Another important bit for documentation. The example in this PR description shows the call to "Bootstrap and join the cluster" as part of the user data. Including this will in fact fail the managed node when NOT using a custom AMI. The above note is from EKS docs. When using the default AMI, the call to bootstrap.sh is merged into the user-data.

We have an eks module wrapping this one, and this is what I used for the LT looks like:

```
data "cloudinit_config" "node_group" {
  gzip          = false
  base64_encode = false

  part {
    content_type = "text/x-shellscript"
    content      = <<-EOT
      yum install -y gnupg
      wget https://inspector-agent.amazonaws.com/linux/latest/install
      wget https://d1wk0tztpsntt1.cloudfront.net/linux/latest/inspector.gpg
      gpg --import inspector.gpg
      wget https://inspector-agent.amazonaws.com/linux/latest/install.sig
      gpg --verify install.sig || exit 1
      sudo bash install
    EOT
  }
}

resource "aws_launch_template" "node_group" {
  for_each = node_groups

  tags = var.tags
  name = "${var.name}-${each.key}-node-group"

  instance_type = lookup(each.value, "instance_type", var.node_group_instance_type)
  user_data     = base64encode(replace(data.cloudinit_config.node_group.rendered, "MIMEBOUNDARY", "//"))

  dynamic "tag_specifications" {
    for_each = ["instance", "volume"]

    content {
      tags          = var.tags
      resource_type = tag_specifications.value
    }
  }
  update_default_version = true
}

module "eks" {
  # Loading from this PR branch for now until launch template configuration is
  # merged: https://github.com/terraform-aws-modules/terraform-aws-eks/pull/997
  source          = "github.com/bahn-x/terraform-aws-eks.git?ref=ca321509"

  // ...

  node_groups = { for key, value in var.node_groups : key => merge({
      instance_type           = var.node_group_instance_type
      min_capacity            = var.node_group_min_capacity
      max_capacity            = var.node_group_max_capacity
      desired_capacity        = var.node_group_min_capacity
      launch_template_id      = aws_launch_template.node_group[key].id
      launch_template_version = aws_launch_template.node_group[key].default_version
    }, value) }
  )
}
```

Replacing the correct multi-part boundary string into the user data seemed to be important or I'd get an error about not using multi-part:
```
user_data     = base64encode(replace(data.cloudinit_config.node_group.rendered, "MIMEBOUNDARY", "//"))
```

Others noted the need to use `iam_role_arn`; wasn't something I ran into. Working without it in my case.

I haven't checked when using a custom AMI, but in my case (where a custom AMI is NOT being used) EKS is creating a new launch template based on the provided one, with the bootstrap script merged into the user data, 

Many things to @philicious for putting this together! I've got it applied to multiple EKS clusters, where I can confirm it's working, and I hope to see it incorporated in some way on a tagged release soon.

philicious commented:
Another thing to note is about using disk/EBS encryption with LaunchTemplates, as also seen in https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1023

I left that out of my initial example but its worth going into the example then too:

If you set 
```
    ebs {
..
      encrypted             = true
      kms_key_id            = var.kms_key_arn
    }
```

you also need to add a key policy to that KMS key, so the cluster-autoscaler can decrypt volumes

```
resource "aws_iam_service_linked_role" "autoscaling" {
  aws_service_name = "autoscaling.amazonaws.com"
  description      = "Default Service-Linked Role enables access to AWS Services and Resources used or managed by Auto Scaling"
}

data "aws_caller_identity" "current" {}

// This policy is required for the EKS KMS, so the cluster is allowed to enc/dec/attach encrypted EBS volumes
data "aws_iam_policy_document" "ebs_decryption" {
  // copy of default KMS policy that lets us manage it
  statement {
    sid    = "Enable IAM User Permissions"
    effect = "Allow"

    principals {
      type        = "AWS"
      identifiers = ["arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"]
    }

    actions = [
      "kms:*"
    ]

    resources = ["*"]
  }

  // required for EKS
  statement {
    sid    = "Allow service-linked role use of the CMK"
    effect = "Allow"

    principals {
      type = "AWS"
      identifiers = [
        "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling", // required for the ASG to manage encrypted volumes for nodes
        module.eks.cluster_iam_role_arn,                                                                                                            // required for the app cluster / persistentvolume-controller to create encrypted PVCs
        module.obs_eks.cluster_iam_role_arn                                                                                                         // required for the obs cluster / persistentvolume-controller to create encrypted PVCs
      ]
    }

    actions = [
      "kms:Encrypt",
      "kms:Decrypt",
      "kms:ReEncrypt*",
      "kms:GenerateDataKey*",
      "kms:DescribeKey"
    ]

    resources = ["*"]
  }

  statement {
    sid    = "Allow attachment of persistent resources"
    effect = "Allow"

    principals {
      type = "AWS"
      identifiers = [
        "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling", // required for the ASG to manage encrypted volumes for nodes
        module.eks.cluster_iam_role_arn,                                                                                                            // required for the app cluster / persistentvolume-controller to create encrypted PVCs
        module.obs_eks.cluster_iam_role_arn                                                                                                         // required for the obs cluster / persistentvolume-controller to create encrypted PVCs
      ]
    }

    actions = [
      "kms:CreateGrant"
    ]

    resources = ["*"]

    condition {
      test     = "Bool"
      variable = "kms:GrantIsForAWSResource"
      values   = ["true"]
    }

  }
}
```

philicious commented:
@barryib any news based on your last fridays working session? I'd like to see this PR merged and I think many other would also be happy. I need to know what I should still add to it so you guys are happy with it :)

huguesalary commented:
It is worth noting there seem to be some kind of bug with EKS right now related to the `instance_types`.

If you specify an `instance_type` (*other than* `t3.medium`) in your `launch_template`, AWS will create a managed node group with the correct `instance_type`, and the EKS Console will correctly report the `instance_type` from the `launch_template`. So far so good.

However, if you try to run `terraform plan`, EKS will always report that the `instance_type` of your node pool is `t3.medium`, even if you set a different `instance_type`. 

This in turn will force the replacement of the node pool, but because it keeps the pet name, you'll run into the error:

```
Error: error creating EKS Node Group (cluster-beta:cluster-beta-foo-growing-civet): ResourceInUseException: NodeGroup already exists with name cluster-beta-foo-growing-civet and cluster name cluster-beta
{
  RespMetadata: {
    StatusCode: 409,
    RequestID: "63e117e5-b237-47d2-837b-abcdefghijklmno"
  },
  ClusterName: "cluster-beta",
  Message_: "NodeGroup already exists with name cluster-beta-foo-growing-civet and cluster name cluster-beta",
  NodegroupName: "cluster-beta-foo-growing-civet"
}
```

If you check with the AWS cli what the `instanceTypes` of your node pool is (`aws eks describe-nodegroup --cluster-name cluster-beta --nodegroup-name cluster-beta-foo-growing-civet`), you will notice that it returns `t3.medium`.

![image](https://user-images.githubusercontent.com/755890/95792624-c2eff480-0c98-11eb-94f2-dfa65b4cb451.png)

philicious commented:
@huguesalary are you sure you havent accidentally also set `instance_type` on the node-group in addition to launchtemplate? 
as I cant reproduce described behavior. i.e. `aws eks describe-nodegroup --cluster-name prod-app-cluster --nodegroup-name prod-app-cluster-initial_group-teaching-mackerel` 

**doesnt** contain `instanceTypes` for me !

huguesalary commented:
I confirm I did not set the instance type, here's my terraform node_group:

```
node_groups = {

    foo = {
      desired_capacity = 3
      max_capacity     = 10
      min_capacity     = 3
      subnets          = module.vpc.private_subnets

      launch_template_id = aws_launch_template.eks_node_with_dockerhub.id
      launch_template_version = aws_launch_template.eks_node_with_dockerhub.latest_version
    }

  }
```

And my launch template:
```
resource "aws_launch_template" "eks_node_with_dockerhub" {
  name = format("%s-%s", "EKS-node-with-dockerhub", var.cluster_name)

  instance_type = "m5.large"
}
```

I also created a new node pool _directly from the aws console_ and get the same result: the aws console shows the instance type from the launch template, but the aws-cli returns `t3.medium`.

huguesalary commented:
Re-creating the cluster from scratch fixed the issue mentioned in comments https://github.com/terraform-aws-modules/terraform-aws-eks/pull/997#issuecomment-707360645 and https://github.com/terraform-aws-modules/terraform-aws-eks/pull/997#issuecomment-707367044

stibi commented:
I just deployed (actually redeployed just the node groups) a couple of EKS cluster and everything is fine. Thanks a lot for this pull request!

How can I help here so it can be merged? Maybe provide an example of the code? Anything else?

vara-bonthu commented:
I am really looking forward for this change to be merged. I have tested this branch and it works fine for me with the launch templates for multiple node groups

barryib commented:
I'm go

> @barryib any news based on your last fridays working session? I'd like to see this PR merged and I think many other would also be happy. I need to know what I should still add to it so you guys are happy with it :)

I'm going to do some tests during the week-end so we can merge it. Thanks in advance for your comprehension.

philicious commented:
@barryib sounds awesome ! I'll allot some weekend time then for adding examples / doc, based on comments in here

myoung34 commented:
Any ETA? Lots of anxious watchers 😆

hawkesnc commented:
This is great! Looking forward to it!

barryib reviewed (changes requested):
Thanks @philicious for your work. Just tested it and sounds ok.

Can you please add an example and update your branch.

philicious commented:
> Thanks @philicious for your work. Just tested it and sounds ok.
> 
> Can you please add an example and update your branch.

on it. will push tonight

philicious commented:
@barryib I now added an example with lots of comments and tested it to be working. imho we would all love to see this merged and hopefully make it soonish into a release *G

@davidalger thx for the hint with user-data and **no** customAMI, added that to example
@Carles-Figuerola thx for the keeper addition

@ everyone-else: thx for testing

jseiser commented:
@philicious 

Thanks for all the work on this, how would one go about using Spot Instances in here?  

Thanks.

RichiCoder1 commented:
@jseiser Unless something has changed, AWS Managed Node Groups does _not_ currently support Spot Instances. Follow https://github.com/aws/containers-roadmap/issues/583 for updates on that feature.

barryib commented on examples/launch_templates_with_managed_node_groups/launchtemplate.tf:
```diff
@@ -0,0 +1,88 @@
+    security_groups             = [module.eks.worker_security_group_id]
+  }
+
+  //image_id      = var.ami_id // if you want to use a custom AMI
+
+  // if you use a custom AMI, you need to supply via user-data, the bootstrap script as EKS DOESNT merge its managed user-data then
+  // you can add more than the minimum code you see in the template, e.g. install SSM agent, see https://github.com/aws/containers-roadmap/issues/593#issuecomment-577181345
+  //
+  // (optionally you can use https://registry.terraform.io/providers/hashicorp/cloudinit/latest/docs/data-sources/cloudinit_config to render the script)
+  # user_data = base64encode(
+  #   data.template_file.launch_template_userdata.rendered,
+  # )
```
Can you please be consistent in your comments and use `#` (replace `//` with `#`) to start all your comments.

barryib commented on examples/launch_templates_with_managed_node_groups/launchtemplate.tf:
```diff
@@ -0,0 +1,88 @@
+  // if you use a custom AMI, you need to supply via user-data, the bootstrap script as EKS DOESNT merge its managed user-data then
+  // you can add more than the minimum code you see in the template, e.g. install SSM agent, see https://github.com/aws/containers-roadmap/issues/593#issuecomment-577181345
+  //
+  // (optionally you can use https://registry.terraform.io/providers/hashicorp/cloudinit/latest/docs/data-sources/cloudinit_config to render the script)
```
I think it's better to use the cloudinit provider. It's more "error proof".

philicious replied:
Ye I can change and try. 
But see this comment where it reads that using cloudinit still needs a little fix for MIME boundary https://github.com/terraform-aws-modules/terraform-aws-eks/pull/997#issuecomment-705286151

barryib replied:
Humm. So you leave it like that. But can you please document it a bit in your example for users who want to use the cloudinit provider. You can even link https://github.com/terraform-aws-modules/terraform-aws-eks/pull/997#issuecomment-705286151 in your example.

barryib replied:
So, It sounds their are errors in AWS docs https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html
![image](https://user-images.githubusercontent.com/550783/97560827-8255e200-19df-11eb-8e96-fcefbba98ea0.png)


FWIW, I opened an issue https://github.com/hashicorp/terraform-provider-cloudinit/issues/5 to add custom boundary option.

barryib commented on examples/launch_templates_with_managed_node_groups/templates/userdata.sh.tpl:
```diff
@@ -0,0 +1,12 @@
+MIME-Version: 1.0
+Content-Type: multipart/mixed; boundary="//"
+
+--//
+Content-Type: text/x-shellscript; charset="us-ascii"
+#!/bin/bash 
+set -xe
+
+# Bootstrap and join the cluster
+/etc/eks/bootstrap.sh --b64-cluster-ca '${cluster_auth_base64}' --apiserver-endpoint '${endpoint}' ${bootstrap_extra_args} --kubelet-extra-args "${kubelet_extra_args}" '${cluster_name}'
+
+--//--
```
I don't test it, can't we use directly the template from the module output `module.eks.workers_user_data` with the cloudinit provider ?

```hcl
data "cloudinit_config" "userdata" {
  base64_encode = true

  part {
    content_type = "text/x-shellscript"
    content      = module.eks.workers_user_data[0]
    filename     = "userdata.sh"
  }
}
```

vara-bonthu replied:
I have tried use this config to test but i am getting the following error
```

variable "eks_user_data" {
value = module.eks.workers_user_data[0]
}

data "cloudinit_config" "userdata" {
  base64_encode = true
  part {
    content_type = "text/x-shellscript"
    content      = var.eks_user_data
    filename     = "userdata.sh"
  }
}
```
Error in Terraform Plan

Error: Invalid index

 ```
 on source/_modules.tf line 246, in module "launch-templates":
 246:   eks_user_data = module.eks.workers_user_data[0]
    |----------------
    | module.eks.workers_user_data is empty tuple

The given key does not identify an element in this collection value.
```

barryib replied:
Just realized that templates are not generated for managed node group. So forget my comment. I'll open another PR to add that feature.

barryib commented on examples/irsa/irsa.tf:
```diff
@@ -1,6 +1,6 @@
 module "iam_assumable_role_admin" {
   source                        = "terraform-aws-modules/iam/aws//modules/iam-assumable-role-with-oidc"
-  version                       = "2.14.0"
+  version                       = "~> v2.0"
```
Is there any reason for this change ? If not, can you please drop it. It was bumped to improve TF 0.13 support. See https://github.com/terraform-aws-modules/terraform-aws-eks/pull/976 for more info.

philicious replied:
Reason was that I used IRSA along the way and noticed it didn't work. Then I changed the version constraint to what I found in aws iam modules readme and then it worked. 
But I can remove from this PR.

barryib replied:
Yep please. Can you please add this change in another PR and explain how this change fixed your problem.

barryib commented on examples/launch_templates_with_managed_node_groups/launchtemplate.tf:
```diff
@@ -0,0 +1,88 @@
+data "template_file" "launch_template_userdata" {
```
Is this necessary ? See my comment on the `userdata.sh.tpl` file.

philicious replied:
I copied from existing code in this module for the unmanaged node groups and there was no cloudinit used hehe. 
So you might also want to make same change there then ?

barryib replied:
Yep. As soon I add the feature to generate userdata template for MNG, I'll update your example accordingly.

barryib reviewed (commented):
Just few minor comments otherwise it's ok.

Hunter-Thompson commented:
> @jseiser Unless something has changed, AWS Managed Node Groups do _not_ currently support Spot Instances. Follow [aws/containers-roadmap#583](https://github.com/aws/containers-roadmap/issues/583) for updates on that feature.

With the support of LaunchTemplates in ManagedNodeGroups we are able to setup spot instances if Im not wrong, I assume this because we are able to set Spot instance options in the LaunchTemplate and then we can create a ManagedNodeGroup with the LaunchTemplate.

Am I wrong? Because clearly we have the option to set Spot instance options in the LaunchTemplate.

davidalger commented:
@Hunter-Thompson Per AWS documentation, the "Request Spot Instances" field on launch templates is prohibited (I haven't tried it to see if the rule is enforced or not) with the following note:

>  	No equivalent is available because Spot Instances are not supported in node groups. 

Reference: https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html

philicious commented:
@barryib i addressed the change requests. Pls have a look 🙂

barryib commented:
Thanks a lot @philicious for your contribution. I'll push a new release during the day.

barryib commented:
[v13.1.0](https://github.com/terraform-aws-modules/terraform-aws-eks/releases/tag/v13.1.0) is now released. Thank you all for your works.

philicious commented:
@barryib thanks for doing such a quick release !! ❤️

yoanisgil commented:
@philicious I just wanted to thank you for your work on this. You're a life saver!

MBalazs90 commented:
I tried this example with custom user-data installing amazon-ssm-agent but Im getting this error

`Error: error waiting for EKS Node Group (test-eks-lt-zPGZQU0N:test-eks-lt-zPGZQU0N-example-polished-leopard) creation: NodeCreationFailure: Instances failed to join the kubernetes cluster. Resource IDs: [i-0b13057dca7c45b3e]`

philicious commented:
@MBalazs90 I have myself rarely but have seen that error before. For different reasons.

For debugging, ssh to the machine and try to CURL the masters URL and also try to run the bootstrap command by hand. 
Thereby making sure that your network setup is fine and allowing that connection.

pre commented:
> @philicious I tried your `launch_templates_with_managed_node_groups` with custom userdata installing amazon-ssm-agent but I'm getting this error
[..]
> `Error: error waiting for EKS Node Group (test-eks-lt-zPGZQU0N:test-eks-lt-zPGZQU0N-example-polished-leopard) creation: NodeCreationFailure: Instances failed to join the kubernetes cluster. Resource IDs: [i-0b13057dca7c45b3e]`
> 
> I am using the default AMI, custom AMI is not enabled

I can replicate this error with custom userdata. 

I have noticed that
* ec2 instances become healthy
* instances join the Kubernetes cluster
* instances do not register as healthy to the managed node group (also visible in the AWS Console > EKS > Compute)

I did spin instances successfully up with managed node groups and the default (implicit) launch template. I inspected the default userdata from the AWS Console > EKS > Compute > MNG > Advanced > Userdata and I noticed that the following labels are present there (in the configuration which works):
```
# Node labels in the default launch template which works
# whatever is in `k8s_labels` +
--node-labels=
eks.amazonaws.com/nodegroup-image=ami-06cfd5b2a2d58e09a,
eks.amazonaws.com/capacityType=ON_DEMAND,
eks.amazonaws.com/sourceLaunchTemplateVersion=5,
eks.amazonaws.com/nodegroup=my-nodegroup-name,
eks.amazonaws.com/sourceLaunchTemplateId=lt-04f5758c3c9507b88
```

So - in order to get managed node groups with launch template WITH custom userdata requirement, you need to MANUALLY fill in the `eks.amazonaws.com/sourceLaunchTemplateVersion`,  `eks.amazonaws.com/nodegroup` and `eks.amazonaws.com/sourceLaunchTemplateId` in order for workers to register to the managed node group. 

I played around with this and found out that those three are the minimum to get nodes register properly. However, the default (implicit) launch template also fills `eks.amazonaws.com/nodegroup-image` and `eks.amazonaws.com/capacityType` so they might serve a real purpose. 

The [examples/launch_templates_with_managed_node_groups/templates/userdata.sh.tpl](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/e4520d2c2e1b1a182d4abb8bdaf032534f5ed89f/examples/launch_templates_with_managed_node_groups/templates/userdata.sh.tpl) provides a reference *but it does not work* without the labels mentioned above.

philicious commented:
I personally only tested with MNG with LT with using a custom AMI, which the userdata template is used for. I never tried userdata w/o a custom AMI. 

However see https://github.com/terraform-aws-modules/terraform-aws-eks/pull/997#issuecomment-705286151 where @davidalger successfully did that. Maybe he has an idea ?

pre commented:
I investigated deeper into this. The problem I was facing is related to the merge of userdata done by EKS Managed Node Groups (MNG).

My problem is that I need to pass custom K8s node-labels to the kubelet. Normally you'd be able to do this by just passing `--kubelet-extra-args '--node-labels xyz=zyz` to `eks-bootstrap.sh` . This is also how aws-eks-terraform normally does it, but this does not work with MNG: https://github.com/terraform-aws-modules/terraform-aws-eks/blob/e4520d2c2e1b1a182d4abb8bdaf032534f5ed89f/templates/userdata.sh.tpl#L7

The problem is that Managed Node Group "merging of userdata" will place a `/etc/eks/bootstrap.sh` command of its own as the last part of the merged userdata. If you call  `/etc/eks/bootstrap.sh` of your own, the script will have only those parameters which _you have provided_. In addition: because the userdata is merged, there will be two calls to `eks-bootstrap.sh`.

The last part of the userdata (provided by MNG) will contain the `eks-boostrap.sh` line with `eks.amazonaws.com` parameters (listed in https://github.com/terraform-aws-modules/terraform-aws-eks/pull/997#issuecomment-728742454)  which will label the node properly in order for the Managed Node Group to determine that the node has successfully joined the cluster. Without these `eks.amazonaws.com` labels, the MNG will display status of "Creation failed". 

## Userdata - update EKS Managed Node Group EC2 instances to the newest AWS Kernel 

* Update to AWS Linux 5.5.x, EKS MNG has only 5.4.x at the moment.
* Reboot after setting the kernel up.
* Join the cluster with `eks-bootstrap.sh` after reboot.
* This is an unpolished example, but it works.

```bash
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="//"

--//
Content-Type: text/x-shellscript; charset="us-ascii"
#!/bin/bash
set -xe

# Install newer Amazon supported kernel
amazon-linux-extras install -y kernel-ng
yum install -y amazon-ssm-agent
yum update -y
 
TOKEN="$(curl -X PUT -H "X-aws-ec2-metadata-token-ttl-seconds: 600" "http://169.254.169.254/latest/api/token")"
INSTANCE_LIFECYCLE="$(curl -H "X-aws-ec2-metadata-token: $TOKEN" -s http://169.254.169.254/latest/meta-data/instance-life-cycle)"
INSTANCE_ID="$(curl -H "X-aws-ec2-metadata-token: $TOKEN" --silent http://169.254.169.254/latest/dynamic/instance-identity/document | jq -r .instanceId -r)"
REGION="$(curl -H "X-aws-ec2-metadata-token: $TOKEN" --silent http://169.254.169.254/latest/dynamic/instance-identity/document | jq -r .region -r)"
LAUNCH_TEMPLATE_VERSION="$(aws ec2 describe-tags --region "$REGION" --filters "Name=resource-id,Values=$INSTANCE_ID" "Name=tag-key,Values=aws:ec2launchtemplate:version" --query 'Tags[0].Value')"
LAUNCH_TEMPLATE_ID="$(aws ec2 describe-tags --region "$REGION" --filters "Name=resource-id,Values=$INSTANCE_ID" "Name=tag-key,Values=aws:ec2launchtemplate:id" --query 'Tags[0].Value')"
NODEGROUP="$(aws ec2 describe-tags --region "$REGION" --filters "Name=resource-id,Values=$INSTANCE_ID" "Name=tag-key,Values=eks:nodegroup-name" --query 'Tags[0].Value')"

# AMI ID is passed by the default MNG launch template, but node joins the cluster without it also. 
# Also as we have just updated the kernel, ami id would need to be queried from somewhere.
# eks.amazonaws.com/nodegroup-image=ami-05cd1e07212dd719a

# TODO: dynamic eks.amazonaws.com/capacityType=ON_DEMAND from INSTANCE_LIFECYCLE
EKS_MNG_LABELS="eks.amazonaws.com/sourceLaunchTemplateVersion=$LAUNCH_TEMPLATE_VERSION,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/sourceLaunchTemplateId=$LAUNCH_TEMPLATE_ID,eks.amazonaws.com/nodegroup=$NODEGROUP"

# https://github.com/awslabs/amazon-eks-ami/blob/0a96824d7b60d0930c846f5d6841d1c10ff411d2/files/bootstrap.sh#L273
K8S_CLUSTER_DNS_IP=172.20.0.10

# Userdata is only executed at the first boot of an EC2 instance.
# Prepare bootstrap instructions which will be executed at the second boot.
cat >/etc/rc.d/rc.local <<EOF
#!/bin/bash
set -xe

# Bootstrap and join the cluster
# https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh
/etc/eks/bootstrap.sh \
  --b64-cluster-ca '${cluster_auth_base64}' \
  --apiserver-endpoint '${endpoint}' \
  --dns-cluster-ip "$K8S_CLUSTER_DNS_IP" \
  ${bootstrap_extra_args} \
  --kubelet-extra-args '--node-labels=${k8s_labels} --node-labels=$EKS_MNG_LABELS' \
  '${cluster_name}'

touch /var/lock/subsys/local
EOF

chmod +x /etc/rc.d/rc.local
systemctl enable rc-local.service

# Start again with the new kernel
reboot

--//--

```

### Launch template:
```hcl
  user_data = base64encode(
    data.template_file.launch_template_userdata_osd.rendered
  )
```

### template_file
```hcl

data "template_file" "launch_template_userdata_osd" {
  template = file("${path.module}/templates/userdata.sh.tpl")

  vars = {
    cluster_name        = var.cluster_name
    endpoint            = module.eks.cluster_endpoint
    cluster_auth_base64 = module.eks.cluster_certificate_authority_data

    bootstrap_extra_args = ""
    k8s_labels   = "node.rdx.net/example-role=example-value"
  }
}
```

I'm over my head now and I'm no longer sure whether the MNG was able to recognize the nodes before I introduced the reboot. However, if you want a newer kernel, the reboot and custom eks-bootstrap.sh is required.

philicious commented:
Wow @pre , That looks like a lot of hassle you had to do to get this working, thanks for sharing !

to me it seems as if using just a custom AMI would be easier: user packer to easily build a custom AMI with the kernel update baked in. Then have the simple userdata from the examples and have a smooth experience with managed node groups and not having to worry about labels yourself

pre commented:
> to me it seems as if using just a custom AMI would be easier

Are you aware of how to install a newer kernel from Amazon Linux Extras as an AMI? The [Amazon FAQ](https://aws.amazon.com/amazon-linux-2/faqs/#Amazon_Linux_Extras) only tells the `amazon-linux-extras install` way. I tried searching using the AWS AMI ID browser in AWS Console but maybe I had wrong keywords, I didn't find any AMI for Amazon Linux Extras.

My goal was not to build an AMI of my own and I definitely did not want a third party AMI. Even though the setup above was a hassle, it now provides a newer Linux 5.5 kernel which is managed by Amazon. 

I tried using the official [AWS Ubuntu](https://cloud-images.ubuntu.com/docs/aws/eks/) image, which also provides Linux 5.5, but it was too different from the Amazon Linux to use without many changes elsewhere. 

TL;DR Do you know how to find an AMI ID for an image which has the latest Amazon Linux Extras with a Linux 5.5 kernel?

pre commented:
PS. If you want to pass custom Kubernetes Node labels  such as `--kubelet-extra-args '--node-labels=${k8s_labels}' ` - you will need custom userdata. Otherwise the nodes are missing the labels which the Managed Node Group expects, and hence, Managed Node Group will fail with "Nodes failed to join the cluster" .

daenney commented:
> encrypt

For the life of me I can't figure out what to do with this. I assume I need to create an AWS IAM policy with that document, but then what? Do I need to attach it to a role? Do I need to do a KMS grant?

It would be really useful to have a complete example.

philicious commented:
@daenney I guess you seen https://github.com/terraform-aws-modules/terraform-aws-eks/tree/master/examples/launch_templates_with_managed_node_groups 
and in particular https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/examples/launch_templates_with_managed_node_groups/launchtemplate.tf#L34-L36 

so the ebs encryption policy should be added as policy to a KMS key as stated, then that KMS key ARN needs to be set in the LT. by that, your cluster node disks will be encrypted and EKS will be able to actually enc/dec them

daenney commented:
Got it. Thanks a ton! Somehow I had missed the `policy` attribute when looking at `aws_kms_key`.

sechunOH commented:
@philicious 
Hello, I'm using managed node-group with launch template.
But, there's something wrong with me. 
If I set no instance_type(in node_group and launch_template), no probs. (instances within that node-group are provisioned with instance type "m4.large". maybe default)
But If I set instance_type in launch_template, shows an error below. (StatusCode: 400)
```
"InvalidRequestException: Cannot specify instance types in launch template and API request"
```

I think this PR may fix it according to above comments, but error occurs.
```
data "template_file" "launch_template_userdata" {
  template = file("${path.module}/templates/userdata.sh.tpl")

  vars = {
    cluster_name        = local.cluster_name
    endpoint            = module.eks.cluster_endpoint
    cluster_auth_base64 = module.eks.cluster_certificate_authority_data

    bootstrap_extra_args = ""
    kubelet_extra_args = ""

    pre_userdata = ""
    additional_userdata = ""
  }
}


module "eks" {
  source          = "terraform-aws-modules/eks/aws"
  version = "15.1.0"

  cluster_name    = local.cluster_name
  cluster_version = "1.19"
  subnets         = module.vpc.private_subnets

  vpc_id          = module.vpc.vpc_id
  enable_irsa     = true

  node_groups = {
    backend = {
      desired_capacity = 2
      max_capacity     = 4
      min_capacity     = 2

      launch_template_id      = aws_launch_template.back.id
      launch_template_version = aws_launch_template.back.default_version
    }
  }

  tags = {
    Environment = "dev"
  }
}

resource "aws_launch_template" "back" {
  name_prefix     = "${local.cluster_name}-"
  description     = "Backend Launch-Template for clusters"
  update_default_version = true

  block_device_mappings {
    device_name = "/dev/xvda"

    ebs {
      volume_size           = 20
      volume_type           = "gp2"
      delete_on_termination = true
    }
  }

  # ebs_optimized = true // some instance types dont support it, so check when changing type

  image_id      = data.aws_ssm_parameter.cluster.value //TODO use our custom AMI
  instance_type = "m4.large"

  monitoring {
    enabled = true
  }

  network_interfaces {
    associate_public_ip_address = false
    delete_on_termination       = true
    security_groups             = [module.eks.worker_security_group_id]
  }

  user_data = base64encode(
    data.template_file.launch_template_userdata.rendered,
  )

  lifecycle {
    create_before_destroy = true
  }  
}
```
There is a little differences with PR's description I think. 
Anyone have the same error as me?
